### PR TITLE
Additional rules for ProGuard

### LINKED ISSUE #2129: Enhancement: retrofit proguard rules
For now the progurad rules in http://square.github.io/retrofit/ won't work correctly if you are using `proguard-android-optimize`. 
We need add this to make it work.
```
-keepclasseswithmembers class * {
    @retrofit2.http.* <methods>;
}
```
For more information please see http://stackoverflow.com/questions/35506111/retrofit-2-not-sending-data-when-proguard-is-enabled.

### PR DESCRIPTION
**Pull request by PromanSEW**
See http://square.github.io/retrofit/
Are these rules still actual?
Also fixes #2129

### CODE REVIEW

**Comment by PromanSEW** [comment]
@swankjesse please review

File: README.md [hunk 1]
```diff
@@ -32,8 +32,15 @@ ProGuard
 
 If you are using ProGuard you might need to add the following options:
 ```
+-dontnote retrofit2.Platform
+-dontwarn retrofit2.Platform$Java8
 -dontwarn okio.**
 -dontwarn javax.annotation.**
+-keepattributes Signature
+-keepattributes Exceptions
+-keepclasseswithmembers class * {
```
Comment (JakeWharton) [inline README.md]: This isn't needed. If you're not referencing the interface it should be removed.

File: README.md [hunk 2]
```diff
@@ -32,8 +32,15 @@ ProGuard
 
 If you are using ProGuard you might need to add the following options:
 ```
+-dontnote retrofit2.Platform
+-dontwarn retrofit2.Platform$Java8
 -dontwarn okio.**
 -dontwarn javax.annotation.**
+-keepattributes Signature
+-keepattributes Exceptions
```
Comment (JakeWharton) [inline README.md]: Why is this needed? We don't throw checked exceptions from inside the proxy.

Comment (PromanSEW) [inline README.md]: See http://square.github.io/retrofit/
ProGuard section. I added lines from these recommendations

Comment (JakeWharton) [inline README.md]: We can remove it from both then.

File: README.md (hunk 1 above)
Comment (PromanSEW) [inline README.md]: If I use these interfaces this block is needed. Maybe note about adding this block is needed here

Comment (JakeWharton) [inline README.md]: We don't use this rule with ProGuard so I'm not convinced it's needed until someone proves otherwise

Comment (PromanSEW) [inline README.md]: OK, I chec again and provide the stacktrace today or tomorrow

Comment (PromanSEW) [inline README.md]: @JakeWharton see provided link in #2338:
https://stackoverflow.com/questions/35506111/retrofit-2-not-sending-data-when-proguard-is-enabled

**Comment by PromanSEW** [comment]
Here is stacktrace without
```
-keepclasseswithmembers class * {
    @retrofit2.http.* <methods>;
}
```
```
HttpCodeException{code=400, response=ApiError{description='Error validating field "path": This field is required.', error='FieldValidationError'}}
at com.yandex.disk.rest.b.b.a(Unknown Source)
at com.yandex.disk.rest.b.b.a(Unknown Source)
at com.yandex.disk.rest.j.a(Unknown Source)
at com.yandex.disk.rest.j.a(Unknown Source)
...
```
Here custom `HttpCodeException` throws when request failed.
```java
private <T> T processResponse(Response<T> response) throws HttpCodeException {
      return response.isSuccessful() ? response.body() : throwHttpCodeException(response);
}
```
I see implementation and here is method from interface:
```java
@GET("/v1/disk/resources/download")
Call<Link> getDownloadLink(@Query("path") String path) throws IOException, ServerIOException;
```
You can see than method has `@Query("path") ` parameter, which is obfuscated while ProGuard processing.
When I added
```
-keepclasseswithmembers class * {
     @retrofit2.http.* <methods>;
}
```
to ProGuard rules, all works fine.

**Comment by JakeWharton** [comment]
Obfuscating the annotation is fine since the reference in Retrofit will
also be updated. The string is a constant so it won't be touched. If the
annotation was removed Retrofit would have thrown an error during
validation.

I'll have to try to reproduce. We don't have this problem internally.

**Comment by priyanshi-gupta** [comment]
With updated JSR-305 annotations, following proguard rules need to be added in ReadMe file:

```
-dontwarn javax.annotation.Nullable
-dontwarn javax.annotation.ParametersAreNonnullByDefault
```

**Comment by NightlyNexus** [comment]
Nullable and ParametersAreNonnullByDefault are already included as `javax.annotation.**`

**Comment by stevestock** [comment]
from #2596...

Please consider keeping [github.com/square/retrofit#proguard](https://github.com/square/retrofit#proguard) and [square.github.io/retrofit](http://square.github.io/retrofit/) ([website code](https://github.com/square/retrofit/blob/master/website/index.html#L176-L189)) in sync:

recommend removing this from the current PR
```
-dontwarn okio.**
-dontwarn javax.annotation.**
```

and adding

> Retrofit uses OKHttp under the hood, so you may want to look at [its ProGuard rules](https://github.com/square/okhttp#proguard) as well.

and then changing the [website code](https://github.com/square/retrofit/blob/master/website/index.html#L176-L189) to match, something like:

> If you are using ProGuard in your project add the following lines to your configuration:
> 
> ```
> # Platform calls Class.forName on types which do not exist on Android to determine platform.
> -dontnote retrofit2.Platform
> # Platform used when running on Java 8 VMs. Will not be used at runtime.
> -dontwarn retrofit2.Platform$Java8
> # ______
>-keepclasseswithmembers class * {
>    @retrofit2.http.* <methods>;
>}
> ```
> Retrofit uses OKHttp under the hood, so you may want to look at [its ProGuard rules](https://github.com/square/okhttp#proguard) as well.
